src/runtime/hexagon_remote/sim_host.cpp

```diff
@@ -0,0 +1,360 @@
+    }
+    return 0;
+}
+
+int send_message(int msg, const std::vector<int> &arguments) {
+    assert(sim);
+
+    HEXAPI_Status status;
+
+    HEX_4u_t remote_msg = 0;
+    status = sim->ReadSymbolValue("rpc_call", &remote_msg);
+    if (status != HEX_STAT_SUCCESS) {
+        printf("HexagonWrapper::ReadSymbolValue(rpcmsg) failed: %d\n", status);
+        return -1;
+    }
+    if (0 != write_memory(remote_msg, &msg, 4)) { return -1; }
+
+    // The arguments are individual numbered variables.
+    for (size_t i = 0; i < arguments.size(); i++) {
+        HEX_4u_t remote_arg = 0;
+        std::string rpc_arg = "rpc_arg" + std::to_string(i);
+        status = sim->ReadSymbolValue(rpc_arg.c_str(), &remote_arg);
+        if (status != HEX_STAT_SUCCESS) {
+            printf("HexagonWrapper::ReadSymbolValue(%s) failed: %d\n", rpc_arg.c_str(), status);
+            return -1;
+        }
+        if (0 != write_memory(remote_arg, &arguments[i], 4)) { return -1; }
+    }
+
+    HEX_4u_t remote_ret = 0;
+    status = sim->ReadSymbolValue("rpc_ret", &remote_ret);
+    if (status != HEX_STAT_SUCCESS) {
+        printf("HexagonWrapper::ReadSymbolValue(rpc_ret) failed: %d\n", status);
+        return -1;
+    }
+
+    HEXAPI_CoreState state;
+    if (msg == Message::Break) {
+        // If we're trying to end the remote simulation, just run
+        // until completion.
+        HEX_4u_t result;
+        state = sim->Run(&result);
+        if (state != HEX_CORE_FINISHED) {
+            printf("HexagonWrapper::Run failed: %d\n", state);
+            return -1;
+        }
+        return 0;
+    } else {
+        // If we want to return and continue simulating, we execute
+        // 1000 cycles at a time, until the remote indicates it has
+        // completed handling the current message.
+        do {
+            HEX_4u_t cycles;
+            state = sim->Step(1000, &cycles);
+            read_memory(&msg, remote_msg, 4);
+            if (msg == Message::None) {
+                HEX_4u_t ret = 0;
+                read_memory(&ret, remote_ret, 4);
+                return ret;
+            }
+        } while (state == HEX_CORE_SUCCESS);
+        printf("HexagonWrapper::StepTime failed: %d\n", state);
+        return -1;
+    }
+}
+
+struct host_buffer {
+    unsigned char *data;
+    int dataLen;
+};
+
+class remote_buffer {
+public:
+    int data;
+    int dataLen;
+
+    remote_buffer() : data(0), dataLen(0) {}
+    remote_buffer(int dataLen) : dataLen(dataLen) {
+        if (dataLen > 0) {
+            data = send_message(Message::Alloc, {dataLen});
+            if (data == 0) {
+                printf("Failed to allocate %d bytes in the Hexagon simulation.\n", dataLen);
+            }
+        } else {
+            data = 0;
+        }
+    }
+    remote_buffer(const void *data, int dataLen) : remote_buffer(dataLen) {
+        if (this->data != 0) {
+            write_memory(this->data, data, dataLen);
```

Comment:
This one in particular is ignored because it's in a constructor, we aren't doing exceptions, and there's nothing to do if it fails. We already print an error.